RHIDP-5490: Configure global header in RHDH (#954)
redhat-developer/red-hat-developers-documentation-rhdh: modules/configuring-the-global-header/proc-customize-rhdh-global-header.adoc

- **`ProfileDropdown`**: Adds a profile dropdown showing the logged-in user's name. By default, it contains two menu items.
- **`MenuItemLink`**: Adds a link item in a dropdown menu. By default, the profile dropdown includes a link to the *Settings* page.
- **`LogoutButton`**: Adds a logout button in the profile dropdown (enabled by default).
- **`CreateDropdown`**: Adds a *Create* dropdown button (disabled by default). The menu items are configurable.
- **`SoftwareTemplatesSection`**: Adds a list of software template links to the *Create* dropdown menu (disabled by default). You must enable `CreateDropdown`.
- **`RegisterAComponentSection`**: Adds a link to the *Register a Component* page in the *Create* dropdown menu (disabled by default). You must enable `CreateDropdown`.

`config.position`::
Specifies the position of the header. Supported values are `above-main-content` and `above-sidebar`.

.Prerequisites
* You must configure the support URL in the `{my-app-config-file}` file to display the *Support* button in the header.
* You must install the notifications plugin to display the *Notifications* button in the header.

.Procedure

> . Copy the default configuration and modify the field values to suit your needs. You can adjust the `priority` value of each header component to control its position. Additionally, you can enable or disable components by adding or removing them from the configuration. To ensure that the remaining header icon buttons align with the end of the header before the profile dropdown button, set `config.props.growFactor` to `1` in the `Spacer` mount point to enable the `Spacer` component. For example:

Reviewer: icon buttons sounds a bit redundant - do you think we should say either icons or buttons?

Author: > icon buttons sounds a bit redundant - do you think we should say either icons or buttons?

Thank you, @linfraze. I was considering `Header icon button`, as it refers to both the header and the icon button. However, since this might seem redundant, I suggest we go with `header button `instead.

+
[source,yaml]
----
- mountPoint: global.header/component
  importName: Spacer
  config:
    priority: 100
    props:
      growFactor: 1
----

> . To use your custom header, you must install it as a dynamic plugin by adding your plugin configuration to your `app-config-dynamic.yaml` file. For example:

Reviewer: I like the clear, simple, straightforward way that this is written - TechDocs add-ons work in a similar way, so I want to copy this language into that doc for consistency :)

+
[source,yaml,subs="+attributes,+quotes"]
----
- package: _<npm_or_oci_package-reference>_
  disabled: false
  pluginConfig:
    dynamicPlugins:
      frontend:
        <package_name>:
          mountPoints:
            - mountPoint: application/header
              importName: _<Application_Header_name>_
              config:
                position: above-main-content
            - mountPoint: global.header/component
              importName: _<Header_component_name>_
              config:
                priority: 100
            - mountPoint: global.header/component
              importName: _<Header_component_name>_
              config:
                priority: 90
----
+
where:

<npm_or_oci_package-reference>:: Specifies the package name.
<Application_Header_name>:: Specifies the name of the application header. For example: `MyHeader`

> <Header_component_name>:: Specifies the name of the header component. For example: `MyHeaderComponent`

Reviewer: Nicely done! Very clear and compliant with guidelines. The only thing to note is that I think we typically use all lowercase in the variables.

[id="customizing-your-product-global-header_{context}"]
= Customizing your {product} global header

You can use dynamic plugins to extend the global header with additional buttons and customize the order and position of icons and features. For more information on enabling dynamic plugins, see link:{installing-and-viewing-dynamic-plugins-url}[{installing-and-viewing-dynamic-plugins-title}].

To extend the functionality of the default global header, include any the following attributes in your global header entry:

`mountPoint`::
Specifies the location of the header. Use `application/header` to specify it as a global header.

`importName`::

> Specifies the component exported by the global header plugin.

Reviewer 2: Hi @christoph-jerolimov , do you think we should give a list of current available header components that are ready for use and explain what it does for each one here? Since we got the feedback from Ben saying there's no documentation of what does the Spacer component do.
```suggestion
Specifies the component exported by the global header plugin. Below is a list of available header components, configurable via the `config` field:

- **`SearchComponent`**: Adds a search bar (enabled by default).
- **`Spacer`**: Adds spacing in the header to position buttons at the end. Useful when `SearchComponent` is disabled.
- **`HeaderIconButton`**: Adds an icon button. A "Create" icon button is enabled by default.
- **`SupportButton`**: Adds a support icon button, allowing users to configure a link to an internal or external page. Enabled by default but requires additional configuration to display.
- **`NotificationButton`**: Adds a notifications icon button to display unread notifications in real time and navigate to the Notifications page. Enabled by default (requires the notifications plugin).
- **`Divider`**: Adds a vertical divider. By default, a divider is placed between the profile dropdown and other header components.
- **`ProfileDropdown`**: Adds a profile dropdown showing the logged-in user’s name. By default, it contains two menu items.
- **`MenuItemLink`**: Adds a link item in a dropdown menu. By default, a link to the "Settings" page is included in the profile dropdown.
- **`LogoutButton`**: Adds a logout button in the profile dropdown (enabled by default).
- **`CreateDropdown`**: Adds a "Create" dropdown button (disabled by default). The menu items are configurable.
- **`SoftwareTemplatesSection`**: Adds a list of software template links to the "Create" dropdown menu (disabled by default). Requires `CreateDropdown` to be enabled.
- **`RegisterAComponentSection`**: Adds a link to the "Register a Component" page in the "Create" dropdown menu (disabled by default). Requires `CreateDropdown` to be enabled.
```

`config.position`::
Specifies the position of the header. Supported values are `above-main-content` and `above-sidebar`.

[NOTE]
====
You can configure several global headers at different positions by adding entries to the `mountPoints` field.
====

.Prerequisites
* You have configured your {product} instance.
* You have configured the support URL in the `{my-app-config-file}` file.

> * You have installed the notification plugin.

Reviewer 2: So is here. This is the package name: backstage-plugin-notifications.
```suggestion
* You have installed the notifications plugin.
```

- **`Divider`**: Adds a vertical divider. By default, a divider appears between the profile dropdown and other header components.
- **`ProfileDropdown`**: Adds a profile dropdown showing the logged-in user's name. By default, it contains two menu items.
- **`MenuItemLink`**: Adds a link item in a dropdown menu. By default, the profile dropdown includes a link to the *Settings* page.
- **`LogoutButton`**: Adds a logout button in the profile dropdown (enabled by default).
- **`CreateDropdown`**: Adds a *Create* dropdown button (disabled by default). The menu items are configurable.
- **`SoftwareTemplatesSection`**: Adds a list of software template links to the *Create* dropdown menu (disabled by default). You must enable `CreateDropdown`.
- **`RegisterAComponentSection`**: Adds a link to the *Register a Component* page in the *Create* dropdown menu (disabled by default). You must enable `CreateDropdown`.

`config.position`::
Specifies the position of the header. Supported values are `above-main-content` and `above-sidebar`.

.Prerequisites
* You have configured the support URL in the `{my-app-config-file}` file.
* You have installed the notifications plugin.

> .Procedure

Author: > I'm trying to show the default configuration before the `Procedure` section. How do you like it this way?

Thank you, @ciiay. I agree that we should add the default configuration before the `Procedure`. I believe having this configuration at the beginning of the section, just before introducing the various attributes in the global header entry, will also help users understand the components of the configuration. Let me know what you think about this.

> . Install your custom header as a dynamic plugin by adding your custom plugin configuration to your `dynamic-plugins.default.yaml` file. For example:
> +
> [source,yaml,subs="+attributes,+quotes"]
> ----
> - package: _<npm_or_oci_package-reference>_
>   disabled: false
>   pluginConfig:
>     dynamicPlugins:
>       frontend:
>         <package_name>:
>           mountPoints:
>             - mountPoint: application/header
>               importName: _<Header_component_name>_
>               config:
>                 position: above-main-content
> ----
> +

Reviewer 2: ```suggestion
.To use your custom header, you need to install it as a dynamic plugin by adding your plugin configuration to your `app-config-dynamic.yaml` file. For example:
+
[source,yaml,subs="+attributes,+quotes"]
----
- package: _<npm_or_oci_package-reference>_
  disabled: false
  pluginConfig:
    dynamicPlugins:
      frontend:
        <package_name>:
          mountPoints:
            - mountPoint: application/header
              importName: _<Application_Header_name>_  # e.g. `MyHeader`
              config:
                position: above-main-content
            - mountPoint: global.header/component
              importName: _<Header_component_name>  # e.g. `MyHeaderComponentA`
              config:
                priority: 100
            - mountPoint: global.header/component
              importName: _<Header_component_name>  # e.g. `MyHeaderComponentB`
              config:
                priority: 90
----
+
```
